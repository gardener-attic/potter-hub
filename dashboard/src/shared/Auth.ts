import Axios, { AxiosResponse } from "axios";
import * as jwt from "jsonwebtoken";
import { K8S_REVERSE_PROXY_URL } from "./Kube";

const AuthTokenKey = "kubeapps_auth_token";
const AuthTokenOIDCKey = "kubeapps_auth_token_oidc";

export const DEFAULT_NAMESPACE = "_all";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string, oidc: boolean) {
    localStorage.setItem(AuthTokenOIDCKey, oidc.toString());
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static usingOIDCToken() {
    return localStorage.getItem(AuthTokenOIDCKey) === "true";
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return ["base64url.bearer.authorization.k8s.io." + token, "binary.k8s.io"];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get(K8S_REVERSE_PROXY_URL + "/", {
        headers: { Authorization: `Bearer ${token}` },
      });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      } else if (res.status === 403) {
        throw new Error("token not authorised for cluster");
      }
      // A 403 authorization error only occurs if the token resulted in
      // successful authentication. We don't make any assumptions over RBAC
      // for the root "/" nonResourceURL or other required authz permissions
      // until operations on those resources are attempted (though we may
      // want to revisit this in the future).
      // In the hub project we want to reject everyone with a 403 error
      // since he can't do anything in the ui anyway. We only have
      // admin or unauthorised users.
      if (res.status !== 403) {
        throw new Error(`${res.status}: ${res.data}`);
      }
    }
  }

  // fetchOIDCToken does a HEAD request to collect the Bearer token
  // from the authorization header if exists
  public static async fetchOIDCToken(): Promise<string | null> {
    try {
      const { headers } = await Axios.head("");
      if (headers && headers.authorization) {
        const tokenMatch = (headers.authorization as string).match(/Bearer\s(.*)/);
        if (tokenMatch) {
          return tokenMatch[1];
        }
      }
    } catch (e) {
      // Unable to retrieve token
    }
    return null;
  }

  // defaultNamespaceFromToken decodes a jwt token to return the k8s service
  // account namespace.
  // TODO(mnelson): until we call jwt.verify on the token during validateToken above
  // we use a default namespace for both invalid tokens and tokens without the expected
  // key.
  public static defaultNamespaceFromToken(token: string) {
    const payload = jwt.decode(token);
    const namespaceKey = "kubernetes.io/serviceaccount/namespace";
    if (!payload || !payload[namespaceKey]) {
      return DEFAULT_NAMESPACE;
    }
    return payload[namespaceKey];
  }
}
