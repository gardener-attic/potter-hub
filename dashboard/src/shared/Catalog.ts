export const allRepositories = {
  value: "_all",
  label: "All repositories",
};
