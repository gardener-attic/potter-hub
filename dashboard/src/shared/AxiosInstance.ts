import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          // Expire the session if we are using OIDC tokens
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.expireSession());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
const a = Axios.create();
a.defaults.timeout = 1800000;

export const axiosWithAuth = a;
