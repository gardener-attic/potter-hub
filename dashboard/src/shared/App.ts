import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
