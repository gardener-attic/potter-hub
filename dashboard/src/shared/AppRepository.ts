import { axiosWithAuth } from "./AxiosInstance";
import { UI_BACKEND_ROOT_URL } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = UI_BACKEND_ROOT_URL;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apprepositories`;
  
  private static getResourceLink(namespace?: string): string {
    return AppRepository.APIEndpoint;
  }

  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/${name}`;
  }
}
