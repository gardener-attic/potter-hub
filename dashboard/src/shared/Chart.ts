import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(URL.api.charts.list(repo));
    return data.data;
  }

  public static async fetchChartVersions(id: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(id),
    );
    return data.data;
  }

  public static async getChartVersion(id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(id, version),
    );
    return data.data;
  }

  public static async getReadme(id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(URL.api.charts.getReadme(id, version));
    return data;
  }

  public static async getValues(id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(URL.api.charts.getValues(id, version));
    return data;
  }

  public static async getSchema(id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(URL.api.charts.getSchema(id, version));
    return data;
  }

  public static async exists(id: string, version: string, repo: string) {
    try {
      await axiosWithAuth.get(URL.api.charts.getVersion(`${repo}/${id}`, version));
    } catch (e) {
      return false;
    }
    return true;
  }

  public static async listWithFilters(name: string, version: string, appVersion: string) {
    const url = `${Chart.APIEndpoint}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/chartsvc/v1";
}
