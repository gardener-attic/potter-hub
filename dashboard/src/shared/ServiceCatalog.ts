import { JSONSchema6 } from "json-schema";
import * as urls from "../shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { IClusterServiceClass } from "./ClusterServiceClass";
import { K8S_REVERSE_PROXY_URL } from "./Kube";
import { IServiceInstance } from "./ServiceInstance";
import { IK8sList, IStatus } from "./types";

export class ServiceCatalog {
  public static async getServiceClasses() {
    return this.getItems<IClusterServiceClass>("clusterserviceclasses");
  }

  public static async getServiceBrokers() {
    return this.getItems<IServiceBroker>("clusterservicebrokers");
  }

  public static async getServicePlans() {
    return this.getItems<IServicePlan>("clusterserviceplans");
  }

  public static async deprovisionInstance(instance: IServiceInstance) {
    const { data } = await axiosWithAuth.delete(
      `${K8S_REVERSE_PROXY_URL}${instance.metadata.selfLink}`,
    );
    return data;
  }

  public static async syncBroker(broker: IServiceBroker) {
    const { data } = await axiosWithAuth.patch<IStatus>(
      urls.api.clusterservicebrokers.sync(broker),
      {
        spec: {
          relistRequests: broker.spec.relistRequests + 1,
        },
      },
      {
        headers: { "Content-Type": "application/merge-patch+json" },
        validateStatus: statusCode => true,
      },
    );
    return data;
  }

  public static async isCatalogInstalled(): Promise<boolean> {
    try {
      const { status } = await axiosWithAuth.get(this.endpoint);
      return status === 200;
    } catch (err) {
      return false;
    }
  }

  public static async getItems<T>(resource: string, namespace?: string): Promise<T[]> {
    const response = await axiosWithAuth.get<IK8sList<T, {}>>(
      this.endpoint + (namespace ? `/namespaces/${namespace}` : "") + `/${resource}`,
    );
    const json = response.data;
    return json.items;
  }

  private static endpoint: string = `${K8S_REVERSE_PROXY_URL}/apis/servicecatalog.k8s.io/v1beta1`;
}
export interface IK8sApiListResponse<T> {
  kind: string;
  apiVersion: string;
  metadata: {
    selfLink: string;
    resourceVersion: string;
  };
  items: T[];
}

export interface ICondition {
  type: string;
  status: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IServiceBroker {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    generation: number;
    creationTimestamp: string;
    finalizers: string[];
  };
  spec: {
    url: string;
    authInfo: any; // Look into
    relistBehavior: string;
    relistDuration: string;
    relistRequests: number;
  };
  status: {
    conditions: ICondition[];
    reconciledGeneration: number;
    lastCatalogRetrievalTime: string;
  };
}

export interface IServicePlan {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    creationTimestamp: string;
  };
  spec: {
    clusterServiceBrokerName: string;
    externalName: string;
    externalID: string;
    description: string;
    externalMetadata?: {
      displayName: string;
      bullets: string[];
    };
    instanceCreateParameterSchema?: JSONSchema6;
    serviceBindingCreateParameterSchema?: JSONSchema6;
    free: boolean;
    clusterServiceClassRef: {
      name: string;
    };
  };
  status: {
    removedFromBrokerCatalog: boolean;
  };
}
