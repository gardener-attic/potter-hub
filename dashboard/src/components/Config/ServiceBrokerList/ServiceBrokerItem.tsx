import * as React from "react";

import { CardFooter } from "../../../components/Card";
import InfoCard from "../../../components/InfoCard";
import azureIcon from "../../../icons/azure.png";
import gcpIcon from "../../../icons/gcp.png";
import placeholder from "../../../img/placeholder.png";
import { IServiceBroker } from "../../../shared/ServiceCatalog";
import SyncButton from "./SyncButton";

interface IServiceBrokerItemProps {
  broker: IServiceBroker;
  sync: (broker: IServiceBroker) => Promise<any>;
}

function getIcon(brokerURL: string) {
  const parser = document.createElement("a");
  parser.href = brokerURL;
  if (parser.hostname.match(/azure/)) {
    return azureIcon;
  } else if (parser.hostname.match(/servicebroker\.googleapis/)) {
    return gcpIcon;
  }
  return placeholder;
}

const ServiceBrokerItem: React.SFC<IServiceBrokerItemProps> = props => {
  const { broker, sync } = props;
  return (
    <InfoCard
      title={broker.metadata.name}
      info={`Last updated ${broker.status.lastCatalogRetrievalTime}`}
      icon={getIcon(broker.spec.url)}
    >
      <CardFooter className="text-c">
        <SyncButton sync={sync} broker={broker} />
      </CardFooter>
    </InfoCard>
  );
};

export default ServiceBrokerItem;
