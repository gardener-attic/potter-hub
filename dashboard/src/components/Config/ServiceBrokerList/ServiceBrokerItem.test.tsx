import { shallow } from "enzyme";
import * as React from "react";

import { IServiceBroker } from "../../../shared/ServiceCatalog";
import ServiceBrokerItem from "./ServiceBrokerItem";

it("renders a card with the service broker", () => {
  const broker = {
    metadata: { name: "wall-street" },
    spec: {
      url: "https://foo.bar",
    },
    status: {
      lastCatalogRetrievalTime: "today",
    },
  } as IServiceBroker;
  const wrapper = shallow(<ServiceBrokerItem broker={broker} sync={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});
