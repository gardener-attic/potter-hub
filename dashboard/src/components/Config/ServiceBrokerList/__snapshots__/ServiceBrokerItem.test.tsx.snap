// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a card with the service broker 1`] = `
<InfoCard
  icon="placeholder.png"
  info="Last updated today"
  title="wall-street"
>
  <CardFooter
    className="text-c"
  >
    <SyncButton
      broker={
        Object {
          "metadata": Object {
            "name": "wall-street",
          },
          "spec": Object {
            "url": "https://foo.bar",
          },
          "status": Object {
            "lastCatalogRetrievalTime": "today",
          },
        }
      }
      sync={[MockFunction]}
    />
  </CardFooter>
</InfoCard>
`;
